hedera-node/hedera-mono-service/src/main/java/com/hedera/services/ledger/accounts/staking/StakeInfoManager.java

```diff
@@ -38,16 +45,51 @@ public StakeInfoManager(final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>>
 
     public void unclaimRewardsForStakeStart(final long nodeId, final long amount) {
         final var info = mutableStakeInfoFor(nodeId);
-        info.increaseUnclaimedStakeRewardStart(amount);
+        if (info != null) {
+            info.increaseUnclaimedStakeRewardStart(amount);
+        }
+    }
+
+    public void prepForManaging(final List<Long> nodeIds) {
+        final var numNodes = nodeIds.size();
+        final var orderedIds = nodeIds.stream().sorted().toList();
+        if (orderedIds.equals(LongStream.range(0, numNodes).boxed().toList())) {
+            cache = new MerkleStakingInfo[numNodes];
+        } else {
+            cache = null;
+        }
     }
 
+    @Nullable
     public MerkleStakingInfo mutableStakeInfoFor(final long nodeId) {
-        // Current node ids are 0, 1, 2, ..., n---if this changes, an array ofc will no longer be a
-        // good cache
-        final var curStakingInfos = stakingInfos.get();
         if (cache == null) {
-            cache = new MerkleStakingInfo[stakingInfos.get().size()];
+            return stakingInfos.get().getForModify(EntityNum.fromLong(nodeId));
```

Comment:
Don't we need to initialize cache here since we call `prepForManaging` only on `RESTART` ?

Reply:
Indeed!